core/cas-server-core-authentication-api/src/main/java/org/apereo/cas/authentication/principal/resolvers/PersonDirectoryPrincipalResolver.java

```diff
@@ -204,12 +204,7 @@ public Principal resolve(final Credential credential, final Optional<Principal>
         attributes.forEach((key, attrValue) -> {
             val values = CollectionUtils.toCollection(attrValue, ArrayList.class);
             LOGGER.debug("Found attribute [{}] with value(s) [{}]", key, values);
-            if (values.size() == 1) {
```

Comment:
I did consider filtering out null values instead but all I wanted to do was fix the crash on null values introduced in 5.3 and not introduce any new behaviour.  Happy to add that change for master though if you prefer.

If you examine the surrounding code this change is ok.  It's only non obvious due to how the code has evolved from 5.3 on.  Right now on master:

- `attrValue` is coerced into an `ArrayList` (always creating a new ArrayList in the process).
- if `attrValue` is a single element list, then the single element is extracted and put into a new `ArrayList` and then added to the attributes map
- else the `ArrayList` is added to the attributes map

So due to the scope of the variables and the fact they are always new collections being created, there are no side effects that can escape the `resolve` method making both branches of the `if` end up with the same state.

Note that in the 5.3 and 6.0 branches the above doesn't hold true so the solution is different.

Reply:
Thanks for the note; makes sense.

I'd feel better if we could remove all null values from the collections. The overall logic should be that if an attribute has no values or only null values, then it should be removed from by resolution engine as a valid attribute. Likewise, if an attribute produces values that are "`hello`, null, `world`", then the null value should also be removed from the list of possible values. I'm not sure if that is the behavior today, and while I think other components do account for the removal/filtering of null too elsewhere but it probably is beneficial to be extra defensive at his layer too.

Reply:
That absolutely makes sense.  I've pushed this for the master change now, 6.0 and 5.3 backports I've left as is because it still feels like a breaking change to me.